pennylane/vqe/vqe.py

```diff
@@ -470,10 +470,29 @@ def __init__(
         """QNodeCollection: The QNodes to be evaluated. Each QNode corresponds to the expectation
         value of each observable term after applying the circuit ansatz."""
 
+        wires = device.wires.tolist()
+
+        tape_mode = qml.tape_mode_active()
+        if tape_mode:
+            try:
+                qml.disable_tape()
+                @qml.qnode(device, interface=interface, diff_method=diff_method, **kwargs)
+                def qnode_for_metric_tensor_in_tape_mode(*qnode_args, **qnode_kwargs):
+                    """The metric tensor cannot currently be calculated in tape-mode QNodes. As a
+                    short-term fix for VQECost, we create a non-tape mode QNode just for
+                    calculation of the metric tensor. In doing so, we reintroduce the same
+                    restrictions of the old QNode but allow users to access new functionality
+                    such as measurement grouping and batch execution of the gradient."""
+                    ansatz(*qnode_args, wires=wires, **qnode_kwargs)
+                    return qml.expval(qml.PauliZ(0))
+                self._qnode_for_metric_tensor_in_tape_mode = qnode_for_metric_tensor_in_tape_mode
+            finally:
+                qml.enable_tape()
```

Comment:
Feels like we need a context manager

Reply:
yeah definitely. While `try`-`finally` is, to all intents and purposes, a low-effort context manager, it would be nice.

Note that `qml.enable_tape()` currently doesn't return anything, so it would be very easy to turn it into a context manager in addition to its current role. For example,

```python

class ExitTapeContext:
    def __enter__(self):
        pass
    def __exit__(self):
        qml.disable_tape()

def enable_tape():
    if _mock_stack:
        return contextlib.nullcontext()

    mocks = [
        mock.patch("pennylane.qnode", qnode),
        mock.patch("pennylane.QNode", QNode),
        mock.patch("pennylane.expval", measure.expval),
        mock.patch("pennylane.var", measure.var),
        mock.patch("pennylane.probs", measure.probs),
        mock.patch("pennylane.sample", measure.sample),
    ]

    with contextlib.ExitStack() as stack:
        for m in mocks:
            stack.enter_context(m)

        _mock_stack.append(stack.pop_all())

    return ExitTapeContext()
```

This way, you could keep using it as before,
```python
qml.enable_tape()
# tape mode
qml.disable_tape()
# non-tape mode
```

or you could use it as a context manager
```python
with qml.enable_tape():
    # tape mode
# non-tape mode
```

Reply:
That's awesome! I'll make a ticket and link to this comment, thanks.